qa/standalone/osd/osd-unexpected-clone.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+#
+# Copyright (C) 2015 Intel <contact@intel.com.com>
+# Copyright (C) 2014, 2015 Red Hat <contact@redhat.com>
+#
+# Author: Xiaoxi Chen <xiaoxi.chen@intel.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Library Public License as published by
+# the Free Software Foundation; either version 2, or (at your option)
+# any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Library Public License for more details.
+#
+
+source $CEPH_ROOT/qa/standalone/ceph-helpers.sh
+
+function run() {
+    local dir=$1
+    shift
+
+    export CEPH_MON="127.0.0.1:7144" # git grep '\<7144\>' : there must be only one
+    export CEPH_ARGS
+    CEPH_ARGS+="--fsid=$(uuidgen) --auth-supported=none "
+    CEPH_ARGS+="--mon-host=$CEPH_MON "
+    CEPH_ARGS+="--osd-objectstore-fuse "
+
+    local funcs=${@:-$(set | sed -n -e 's/^\(TEST_[0-9a-z_]*\) .*/\1/p')}
+    for func in $funcs ; do
+        setup $dir || return 1
+        $func $dir || return 1
+        teardown $dir || return 1
+    done
+}
+
+function TEST_recover_unexpected() {
+    local dir=$1
+
+    run_mon $dir a || return 1
+    run_mgr $dir x || return 1
+    run_osd $dir 0 || return 1
+    run_osd $dir 1 || return 1
+    run_osd $dir 2 || return 1
+
+    ceph osd pool create foo 1
+    rados -p foo put foo /etc/passwd
+    rados -p foo mksnap snap
+    rados -p foo put foo /etc/motd
+
+    find $dir
+    cp $dir/1/fuse/1.0_head/all/#1\:602f83fe\:\:\:foo\:1#/attr/_ _
+    cp $dir/1/fuse/1.0_head/all/#1\:602f83fe\:\:\:foo\:1#/data data
+
+    rados -p foo rmsnap snap
+
+    sleep 5
+
+    primary=1
+    mkdir $dir/$primary/fuse/1.0_head/all/#1\:602f83fe\:\:\:foo\:1#
+    cat data > $dir/$primary/fuse/1.0_head/all/#1\:602f83fe\:\:\:foo\:1#/data
+    cat _ > $dir/$primary/fuse/1.0_head/all/#1\:602f83fe\:\:\:foo\:1#/attr/_
```

Comment:
local osd=$(get_primary foo foo)
objectstore_tool $dir $osd foo set-bytes data
objectstore_tool $dir $osd foo set-attr _  _

Reply:
will set-bytes create the object if it doesn't exist?  how do I specify the snapid portion?

Reply:
JSON="$(objectstore_tool $dir $osd --op list foo | grep "snapid":1)"
objectstore_tool $dir $osd $JSON set-bytes data
objectstore_tool $dir $osd $JSON set-attr _ _

Reply:
Using objectstore_tool() can be slow because it kills and restarts the OSD every time.  It could also break the test if recovery occurs during this process.  If that happens use kill_daemon() to kill the OSD(s) first, then use ceph-objectstore-tool directly and run_osd() to restart it afterwards.